modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/io/RowVersionDataIo.java

```diff
@@ -50,46 +52,102 @@ protected RowVersionDataIo(int ver) {
         assertPageType(pageAddr);
 
         long addr = pageAddr + dataOff;
 
         putShort(addr, 0, (short) payloadSize);
         addr += Short.BYTES;
 
-        addr += HybridTimestamps.writeTimestampToMemory(addr, 0, row.timestamp());
+        addr += HybridTimestamps.writeTimestampToMemory(addr, 0, rowVersion.timestamp());
 
-        addr += writePartitionless(addr, row.nextLink());
+        addr += writePartitionless(addr, rowVersion.nextLink());
 
-        putInt(addr, 0, row.valueSize());
+        putInt(addr, 0, rowVersion.valueSize());
         addr += Integer.BYTES;
 
-        putByteBuffer(addr, 0, row.value());
+        BinaryRow row = rowVersion.value();
+
+        if (row != null) {
+            // Write schema version in LE order.
+            putByte(addr, 0, (byte) row.schemaVersion());
+            putByte(addr, 1, (byte) (row.schemaVersion() >> Byte.SIZE));
+            putByte(addr, 2, (byte) (row.hasValue() ? 1 : 0));
+            putByteBuffer(addr, 3, row.tupleSlice());
+        }
     }
 
     @Override
-    protected void writeFragmentData(RowVersion row, ByteBuffer pageBuf, int rowOff, int payloadSize) {
+    protected void writeFragmentData(RowVersion rowVersion, ByteBuffer pageBuf, int rowOff, int payloadSize) {
         assertPageType(pageBuf);
 
         if (rowOff == 0) {
             // first fragment
-            assert row.headerSize() <= payloadSize : "Header must entirely fit in the first fragment, but header size is "
-                    + row.headerSize() + " and payload size is " + payloadSize;
+            assert rowVersion.headerSize() <= payloadSize : "Header must entirely fit in the first fragment, but header size is "
+                    + rowVersion.headerSize() + " and payload size is " + payloadSize;
 
-            HybridTimestamps.writeTimestampToBuffer(pageBuf, row.timestamp());
+            HybridTimestamps.writeTimestampToBuffer(pageBuf, rowVersion.timestamp());
 
-            PartitionlessLinks.writeToBuffer(pageBuf, row.nextLink());
+            PartitionlessLinks.writeToBuffer(pageBuf, rowVersion.nextLink());
 
-            pageBuf.putInt(row.valueSize());
+            pageBuf.putInt(rowVersion.valueSize());
 
-            putValueBufferIntoPage(pageBuf, row.value(), 0, payloadSize - row.headerSize());
+            payloadSize -= rowVersion.headerSize();
         } else {
             // non-first fragment
-            assert rowOff >= row.headerSize();
+            assert rowOff >= rowVersion.headerSize();
+
+            rowOff -= rowVersion.headerSize();
+        }
+
+        BinaryRow row = rowVersion.value();
+
+        if (payloadSize == 0 || row == null) {
+            return;
+        }
+
+        writeFragmentedBinaryRow(row, pageBuf, rowOff, payloadSize);
+    }
+
+    private void writeFragmentedBinaryRow(BinaryRow row, ByteBuffer pageBuf, int offset, int payloadSize) {
```

Comment:
I think you made it overly complicated. You could write "version" right before `pageBuf.putInt(rowVersion.valueSize());`, for example, and then all you had to do is just leave the old code, basically (considering that you'll make sure to store version before size in other places)